modules/api/pkg/ee/clusterbackup/backupstoragelocation/handler.go

```diff
@@ -0,0 +1,458 @@
+			}
+			if bsl.Spec.Credential.Key == "" {
+				bsl.Spec.Credential.Key = defaultCloudCredentialsSecretKeyName
+			}
+			secret = &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      secretRefName,
+					Namespace: clusterbackup.UserClusterBackupNamespace,
+					Labels:    getBSLLabels(bslFullName, projectID, clusterID),
+				},
+				Data: map[string][]byte{
+					bsl.Spec.Credential.Key: cloudCreds,
+				},
+			}
+		}
+	}
+
+	bsl.ObjectMeta = metav1.ObjectMeta{
+		Name:      bslFullName,
+		Namespace: clusterbackup.UserClusterBackupNamespace,
+		Labels:    getBSLLabels(bslName, projectID, clusterID),
+	}
+
+	if !existingSecret {
+		bsl.Spec.Credential = &corev1.SecretKeySelector{
+			LocalObjectReference: corev1.LocalObjectReference{
+				Name: secret.Name,
+			},
+			Key: credentialsSecretKeyName,
+		}
+	}
+
+	if err := client.Create(ctx, bsl); err != nil {
+		return nil, common.KubernetesErrorToHTTPError(err)
+	}
+
+	ownerReferences := metav1.OwnerReference{
+		APIVersion: velerov1.SchemeGroupVersion.String(),
+		Kind:       bslKind,
+		Name:       bslFullName,
+		UID:        bsl.UID,
+	}
+	if !existingSecret {
+		secret.ObjectMeta.OwnerReferences = append(secret.ObjectMeta.OwnerReferences, ownerReferences)
+		if err := client.Create(ctx, secret); err != nil {
+			return nil, common.KubernetesErrorToHTTPError(err)
+		}
+	} else {
+		ownerRefExist := false
+		for _, ownerRef := range secret.ObjectMeta.OwnerReferences {
+			if ownerRef.UID == bsl.UID {
+				ownerRefExist = true
+				break
+			}
+		}
+		if !ownerRefExist {
+			updatedSecret := secret.DeepCopy()
+			updatedSecret.ObjectMeta.OwnerReferences = append(updatedSecret.ObjectMeta.OwnerReferences, ownerReferences)
+			if err := client.Patch(ctx, updatedSecret, ctrlruntimeclient.MergeFrom(secret)); err != nil {
+				return nil, common.KubernetesErrorToHTTPError(err)
+			}
+		}
+	}
+	return bsl, nil
+}
+
+func createHashFromBslSpec(bslSpec velerov1.BackupStorageLocationSpec) string {
+	s3Url := bslSpec.Config[bslS3Url]
+	region := bslSpec.Config[bslRegion]
+	hash := sha256.Sum256([]byte(s3Url + region + bslSpec.ObjectStorage.Bucket + bslSpec.ObjectStorage.Prefix))
+	hashStr := hex.EncodeToString(hash[:])[:12]
+	hashStr = strings.ToLower(hashStr)
+	hashStr = strings.ReplaceAll(hashStr, "_", "-")
+	if len(hashStr) > 6 {
+		hashStr = hashStr[:6]
+	}
+	return hashStr
+}
+
+func validateBsl(ctx context.Context, client ctrlruntimeclient.Client, bslSpec velerov1.BackupStorageLocationSpec) error {
+	bslList := &velerov1.BackupStorageLocationList{}
+	if err := client.List(ctx, bslList, ctrlruntimeclient.InNamespace(clusterbackup.UserClusterBackupNamespace)); err != nil {
+		return common.KubernetesErrorToHTTPError(err)
+	}
+
+	if len(bslList.Items) == 0 {
+		return nil
+	}
+	for _, bsl := range bslList.Items {
+		if isBslSpecIdentical(bsl.Spec, bslSpec) {
+			return fmt.Errorf("bsl with same config already present")
+		}
+	}
+	return nil
+}
```

Comment:
Unfortunately, this looks like a wrong approach to me; the config of multiple BSLs can be the same, and that's not something we'd like to control. If you want to check if a BSL already exists than name, CBSL name should be enough for the check.

Reply:
When you create a BSL, store the CBSL name in its labels or annotations